### PR TITLE
Otimize/atomic write file

### PR DESCRIPTION
**Pull request by bxy4543**
automic write file
#78 
### Describe how you did it

test report：https://github.com/alibaba/sealer/runs/3005104680?check_suite_focus=true

### CODE REVIEW

File: utils/fswriters.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+func (a *atomicFileWriter) Write(p []byte) (n int, err error) {
+	n, err = a.f.Write(p)
+	if err != nil {
+		a.writeErr = err
+	}
+	return n, err
+}
+
+func (a *atomicFileWriter) Close() (err error) {
```
Comment (justadogistaken) [inline utils/fswriters.go:24]: close will commit the file, right? how about "close", I guess Close is not recommended using by caller

File: utils/file.go [hunk 2]
```diff
@@ -71,6 +71,19 @@ func MkTmpdir() (string, error) {
...
 }
 
+func MkTmpFile(path string) (*os.File, error) {
+	if path == "" {
+		path = os.TempDir()
+	}
+	if _, err := os.Stat(filepath.Dir(path)); err != nil {
+		return nil, err
+	}
+	tempFile, err := ioutil.TempFile(filepath.Dir(path), ".tmp-"+filepath.Base(path))
```
Comment (justadogistaken) [inline utils/file.go:81]: How about we define a static path under /var/lib/sealer/tmp, which could help us clean garbage easier.

File: utils/fswriters.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		return err
+	}
+	if err := a.f.Close(); err != nil {
+		return err
+	}
+	if err := os.Chmod(a.f.Name(), a.perm); err != nil {
+		return err
+	}
+	if a.writeErr == nil {
+		return os.Rename(a.f.Name(), a.path)
```
Comment (justadogistaken) [inline utils/fswriters.go:40]: What could happen if a.path exists already? os.Rename can overwrite it?
return err or overwrite the file? which is better?

Comment (bxy4543) [reply utils/fswriters.go:40]: If newpath already exists and is not a directory, Rename replaces it

**Comment by justadogistaken** [comment]
@bxy4543 LGTM, Have you tested the feature already?

**Review by justadogistaken — DISMISSED** [review]

**Comment by justadogistaken** [comment]
@bxy4543 LGTM, you can merge this pr if you've tested it already.

**Comment by bxy4543** [comment]
> @bxy4543 LGTM，如果你已经测试过了，你可以合并这个公关。

OK

File: utils/file.go [hunk 4]
```diff
@@ -67,10 +67,22 @@ func Mkdir(dirName string) error {
...
-	tempDir := filepath.Join(os.TempDir(), GenUniqueID(32))
+	tempDir, err := ioutil.TempDir(common.DefaultTmpDir, ".DTmp-")
+	if err != nil {
+		return "", err
+	}
 	return tempDir, os.MkdirAll(tempDir, os.ModePerm)
 }
 
+func MkTmpFile() (*os.File, error) {
+	if !IsDir(common.DefaultTmpDir) {
```
Comment (justadogistaken) [inline utils/file.go:78]: You can mkdir for DefaultTmpDir at boot module to ensure this dir existence and its type

File: utils/fswriters.go (hunk 3 above)
Comment (justadogistaken) [reply utils/fswriters.go:40]: If my target is "/test/sub/file", but dir "/test/sub" not exists, what will happen?

File: utils/fswriters.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+func AtomicWriteFile(filepath string, data []byte, perm os.FileMode) error {
+	f, err := newAtomicFileWriter(filepath, perm)
+	if err != nil {
+		return err
+	}
+	n, err := f.write(data)
+	if err == nil && n < len(data) {
+		f.writeErr = io.ErrShortWrite
+	}
+	if closeErr := f.close(); closeErr != nil {
```
Comment (justadogistaken) [inline utils/fswriters.go:62]: why not just return f.close()? I do see the difference.

File: utils/fswriters.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+func AtomicWriteFile(filepath string, data []byte, perm os.FileMode) error {
+	f, err := newAtomicFileWriter(filepath, perm)
+	if err != nil {
+		return err
+	}
+	n, err := f.write(data)
+	if err == nil && n < len(data) {
+		f.writeErr = io.ErrShortWrite
+	}
```
Comment (justadogistaken) [inline utils/fswriters.go:61]: if err == nil && n < len(data) {
		f.writeErr = io.ErrShortWrite
}
we can put this part into func f.write, couldn't we?

File: utils/fswriters.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"io"
+	"os"
+)
+
+type atomicFileWriter struct {
+	f        *os.File
+	path     string
+	writeErr error
```
Comment (justadogistaken) [inline utils/fswriters.go:11]: do we really need writeErr?

File: utils/fswriters.go (hunk 6 above)
Comment (bxy4543) [reply utils/fswriters.go:61]: yes

File: utils/fswriters.go (hunk 7 above)
Comment (bxy4543) [reply utils/fswriters.go:11]: I'll delete it

File: utils/file.go (hunk 4 above)
Comment (bxy4543) [reply utils/file.go:78]: OK

**Review by justadogistaken — APPROVED** [review]

**Pull request merged**